pageserver/src/layered_repository/timeline.rs

```diff
@@ -634,6 +644,58 @@ impl LayeredTimeline {
         Ok(())
     }
 
+    /// (Re-)calculate the logical size of the database at the latest LSN.
+    ///
+    /// This can be a slow operation.
+    pub fn init_logical_size(&self) -> Result<()> {
+        // Try a fast-path first:
+        // Copy logical size from ancestor timeline if there has been no changes on this
+        // branch, and no changes on the ancestor branch since the branch point.
+        if self.get_ancestor_lsn() == self.get_last_record_lsn() && self.ancestor_timeline.is_some()
+        {
+            let ancestor = self.get_ancestor_timeline()?;
+            let ancestor_logical_size = ancestor.get_current_logical_size();
+            // Check LSN after getting logical size to exclude race condition
+            // when ancestor timeline is concurrently updated.
+            //
+            // Logical size 0 means that it was not initialized, so don't believe that.
+            if ancestor_logical_size != 0 && ancestor.get_last_record_lsn() == self.ancestor_lsn {
```

Comment:
Look like with this change, it goes back to the copy-on-loading approach in https://github.com/neondatabase/neon/pull/2117. One reason we did another PR (https://github.com/neondatabase/neon/pull/2139) was because it might be better to copy-on-branch-creation as we know that the statement `ancestor.get_last_record_lsn() == self.ancestor_lsn` holds during the branch creation. But that's not true when loading the child because parent branch can be modified between.

I'm fine with either approach as both will improve the branch creation time. I still think the one we did in #2139 will bring more gain. 

cc @knizhnik

Reply:
Hmm, create_timeline calls `repo.branch_timeline`, and immediately after that, it calls `tenant_mgr::get_local_timeline_with_load`, which calls `load_local_timeline`, which calls `init_logical_size`. Admittedly, there are some semi-expensive operations in between, like writing out the metadata file. But even with the previous coding, if new data is inserted to the parent branch at the same time, the optimization might not work, so this just makes the window slightly larger. I'm ok with that.

In the long run, I think lazy initialization is the way to go. There's no strong reason we couldn't allow the user to use the branch, before the logical timeline size has been calculated.

Reply:
> In the long run, I think lazy initialization is the way to go. There's no strong reason we couldn't allow the user to use the branch, before the logical timeline size has been calculated.

That makes sense. Created https://github.com/neondatabase/neon/issues/2167 for that